File: app/AgentControlMac/Sources/Shared/Core/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case invalidURL
    case httpError(Int, String)

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid URL"
        case .httpError(let code, let body): return "HTTP \(code): \(body)"
        }
    }
}

final class APIClient {
    private(set) var baseURL = ""
    private(set) var token = ""
    private var skipTLSVerify = false
    private var customSession: URLSession?
    private var tlsBypassDelegate: TLSBypassDelegate?

    private var session: URLSession {
        if skipTLSVerify {
            if customSession == nil {
                tlsBypassDelegate = TLSBypassDelegate()
                customSession = URLSession(configuration: .default, delegate: tlsBypassDelegate, delegateQueue: nil)
            }
            return customSession!
        }
        customSession?.invalidateAndCancel()
        customSession = nil
        tlsBypassDelegate = nil
        return URLSession.shared
    }

    func configure(baseURL: String, token: String, skipTLSVerify: Bool = false) {
        self.baseURL = baseURL.hasSuffix("/") ? String(baseURL.dropLast()) : baseURL
        self.token = token
        if !skipTLSVerify {
            customSession?.invalidateAndCancel()
            customSession = nil
            tlsBypassDelegate = nil
        }
        self.skipTLSVerify = skipTLSVerify
    }

    /// One-off connectivity check with given config (does not change instance state).
    static func checkConnection(baseURL: String, token: String, skipTLSVerify: Bool) async throws {
        let base = baseURL.hasSuffix("/") ? String(baseURL.dropLast()) : baseURL
        guard let url = URL(string: base + "/api/servers") else { throw APIError.invalidURL }
        var req = URLRequest(url: url)
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let session: URLSession
        if skipTLSVerify {
            let delegate = TLSBypassDelegate()
            session = URLSession(configuration: .default, delegate: delegate, delegateQueue: nil)
        } else {
            session = URLSession.shared
        }
        let (data, response) = try await session.data(for: req)
        if skipTLSVerify { session.invalidateAndCancel() }
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidURL }
        if !(200...299).contains(http.statusCode) {
            throw APIError.httpError(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
    }

    // MARK: - REST endpoints

    func fetchServers() async throws -> [Server] {
        let data = try await request("/api/servers")
        return try JSONDecoder().decode(ServersResponse.self, from: data).servers
    }

    func fetchSessions(serverID: String?) async throws -> [Session] {
        var path = "/api/sessions"
        if let sid = serverID, !sid.isEmpty {
            path += "?server_id=\(sid)"
        }
        let data = try await request(path)
        return try JSONDecoder().decode(SessionsResponse.self, from: data).sessions
    }

    func createSession(
        serverID: String, cwd: String, resumeID: String?,
        env: [String: String], cols: Int, rows: Int
    ) async throws -> Session {
        var body: [String: Any] = [
            "server_id": serverID, "cwd": cwd,
            "env": env, "cols": cols, "rows": rows,
        ]
        if let rid = resumeID, !rid.isEmpty { body["resume_id"] = rid }
        let jsonData = try JSONSerialization.data(withJSONObject: body)
        let data = try await request("/api/sessions", method: "POST", body: jsonData)
        return try JSONDecoder().decode(Session.self, from: data)
    }

    func stopSession(_ sessionID: String, graceMS: Int = 4000, killAfterMS: Int = 9000) async throws {
        let body = try JSONSerialization.data(withJSONObject: [
            "grace_ms": graceMS, "kill_after_ms": killAfterMS,
        ])
        _ = try await request("/api/sessions/\(sessionID)/stop", method: "POST", body: body)
    }

    func fetchEvents(_ sessionID: String) async throws -> [[String: Any]] {
        let data = try await request("/api/sessions/\(sessionID)/events")
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let events = json["events"] as? [[String: Any]] else { return [] }
        return events
    }

    // MARK: - Internal

    private func request(_ path: String, method: String = "GET", body: Data? = nil) async throws -> Data {
        guard let url = URL(string: baseURL + path) else { throw APIError.invalidURL }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        if let body {
            req.httpBody = body
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        let (data, response) = try await session.data(for: req)
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw APIError.httpError(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
        return data
    }
}

File: app/AgentControlWin/Core/ApiClient.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Net.Security;

namespace AgentControlWin.Core;

public class ApiException : Exception
{
    public int StatusCode { get; }
    public ApiException(int code, string message) : base($"HTTP {code}: {message}") => StatusCode = code;
}

public class ApiClient
{
    private string _baseUrl = "";
    private string _token = "";
    private bool _skipTlsVerify;
    private HttpClient? _httpClient;

    private HttpClient HttpClient
    {
        get
        {
            if (_httpClient != null) return _httpClient;
            var handler = new HttpClientHandler();
            if (_skipTlsVerify)
            {
                handler.ServerCertificateCustomValidationCallback =
                    (_, _, _, _) => true;
            }
            _httpClient = new HttpClient(handler);
            _httpClient.DefaultRequestHeaders.Authorization =
                new AuthenticationHeaderValue("Bearer", _token);
            return _httpClient;
        }
    }

    public void Configure(string baseUrl, string token, bool skipTlsVerify = false)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        _token = token;
        _skipTlsVerify = skipTlsVerify;
        // Reset client to pick up new settings
        _httpClient?.Dispose();
        _httpClient = null;
    }

    public static async Task CheckConnectionAsync(string baseUrl, string token, bool skipTlsVerify)
    {
        var url = baseUrl.TrimEnd('/') + "/api/servers";
        var handler = new HttpClientHandler();
        if (skipTlsVerify)
            handler.ServerCertificateCustomValidationCallback = (_, _, _, _) => true;
        using var client = new HttpClient(handler);
        client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
        var resp = await client.GetAsync(url);
        if (!resp.IsSuccessStatusCode)
        {
            var body = await resp.Content.ReadAsStringAsync();
            throw new ApiException((int)resp.StatusCode, body);
        }
    }

    public async Task<List<Server>> GetServersAsync()
    {
        var data = await RequestAsync("/api/servers");
        var resp = JsonSerializer.Deserialize<ServersResponse>(data);
        return resp?.Servers ?? [];
    }

    public async Task<List<Session>> GetSessionsAsync(string? serverId)
    {
        var path = "/api/sessions";
        if (!string.IsNullOrEmpty(serverId))
            path += $"?server_id={Uri.EscapeDataString(serverId)}";
        var data = await RequestAsync(path);
        var resp = JsonSerializer.Deserialize<SessionsResponse>(data);
        return resp?.Sessions ?? [];
    }

    public async Task<Session> CreateSessionAsync(
        string serverId, string cwd, string? sessionId,
        Dictionary<string, string> env, SessionType sessionType,
        int cols = 120, int rows = 30)
    {
        var body = new Dictionary<string, object>
        {
            ["server_id"] = serverId,
            ["session_type"] = sessionType == SessionType.Chat ? "chat" : "pty",
            ["cwd"] = cwd,
            ["env"] = env,
        };
        if (sessionType == SessionType.Pty)
        {
            body["cols"] = cols;
            body["rows"] = rows;
        }
        if (!string.IsNullOrEmpty(sessionId))
            body["session_id"] = sessionId!.ToLower();
        var json = JsonSerializer.Serialize(body);
        var data = await RequestAsync("/api/sessions", "POST", json);
        return JsonSerializer.Deserialize<Session>(data) ?? throw new Exception("Invalid session response");
    }

    public async Task StopSessionAsync(string sessionId, int graceMs = 4000, int killAfterMs = 9000)
    {
        var body = JsonSerializer.Serialize(new { grace_ms = graceMs, kill_after_ms = killAfterMs });
        await RequestAsync($"/api/sessions/{sessionId}/stop", "POST", body);
    }

    public async Task DeleteSessionAsync(string sessionId)
    {
        await RequestAsync($"/api/sessions/{sessionId}", "DELETE");
    }

    public async Task<List<ChatMessage>> GetChatHistoryAsync(string sessionId)
    {
        var data = await RequestAsync($"/api/sessions/{sessionId}/chat");
        var resp = JsonSerializer.Deserialize<ChatHistoryResponse>(data);
        return resp?.Messages ?? [];
    }

    public async Task<Session> SwitchSessionAsync(
        string sessionId, SessionType sessionType,
        Dictionary<string, string>? env = null,
        int cols = 120, int rows = 30)
    {
        var body = new Dictionary<string, object>
        {
            ["session_type"] = sessionType == SessionType.Chat ? "chat" : "pty",
            ["env"] = env ?? [],
        };
        if (sessionType == SessionType.Pty)
        {
            body["cols"] = cols;
            body["rows"] = rows;
        }
        var json = JsonSerializer.Serialize(body);
        var data = await RequestAsync($"/api/sessions/{sessionId}/switch", "POST", json);
        return JsonSerializer.Deserialize<Session>(data) ?? throw new Exception("Invalid session response");
    }

    private async Task<string> RequestAsync(string path, string method = "GET", string? jsonBody = null)
    {
        var url = _baseUrl + path;
        using var request = new HttpRequestMessage(new HttpMethod(method), url);
        if (jsonBody != null)
        {
            request.Content = new StringContent(jsonBody, Encoding.UTF8, "application/json");
        }
        var response = await HttpClient.SendAsync(request);
        var body = await response.Content.ReadAsStringAsync();
        if (!response.IsSuccessStatusCode)
            throw new ApiException((int)response.StatusCode, body);
        return body;
    }
}
